# Functions

> Eric has \$5, Lilly has \$13, and Rahul has \$22. Rahul gives \$2 to his friend, Ashley. Then, he lets Eric borrow all the money he has left. Eric spends all his money. Then, Lilly gives half her money to Eric. Eric gives Rahul back the same amount that he borrowed earlier. How much money do Eric, Lilly, and Rahul have?

In [20]:
#Starting money
eric_money = 5.0
lilly_money = 13.0
rahul_money = 7.0

#Rahul gives $2 to Ashley
rahul_to_ashley = 2.0
rahul_money = rahul_money - rahul_to_ashley

#Eric borrows all of Rahul's money
rahul_to_eric = rahul_money
rahul_money = rahul_money - rahul_to_eric
eric_money = eric_money + rahul_to_eric

#Eric spends all his money
eric_money = eric_money - eric_money

#Lilly gives half her money to Eric
lilly_to_eric = lilly_money / 2.0
lilly_money = lilly_money - lilly_to_eric
eric_money = eric_money + lilly_to_eric

#Eric gives Rahul back the same amount he borrowed earlier
eric_to_rahul = rahul_to_eric
eric_money = eric_money - eric_to_rahul
rahul_money = rahul_money + eric_to_rahul

#How much money do Eric, Lilly, and Rahul have?
print "Eric has $" + str(eric_money)
print "Lilly has $" + str(lilly_money)
print "Rahul has $" + str(rahul_money)

Eric has $1.5
Lilly has $6.5
Rahul has $5.0


This is great and answers the question, but it's kind of a lot of code. We are going to use a <b>function</b> to simplify our code. While a <b>variable</b> allows you to reuse a value, a <b>function</b> allows you to reuse one or more lines of code. Below is an example function.

In [18]:
def give(giver_amount, taker_amount, amount_to_give):
    giver_amount = giver_amount - amount_to_give
    taker_amount = taker_amount + amount_to_give
    return giver_amount, taker_amount

We aren't going to worry about how to make our own functions right now, but here is some language we will use to describe this function:

1. The function is named <b>give</b>
2. The function takes three <b>arguments</b>: giver_amount, taker_amount, and amount_to_give IN THAT ORDER
3. The function returns two <b>values</b>: giver_amount and taker_amount IN THAT ORDER

This tells us everything we need to know to use, or <b>call</b>, this function. We need to give it three pieces of information, or <b>arguments</b>, and it will gives us back two values. Below is an example.

In [17]:
eric_money = 12.0
lilly_money = 10.0
print "Eric has $" + str(eric_money)
print "Lilly has $" + str(lilly_money)

print "Something happened"
eric_money, lilly_money = give(eric_money, lilly_money, 6.5)

print "Eric has $" + str(eric_money)
print "Lilly has $" + str(lilly_money)

Eric has $12.0
Lilly has $10.0
Something happened
Eric has $5.5
Lilly has $16.5


<p>
As we can see, Eric gave Lilly the $6.50 he borrowed from her in the last problem. More importantly, we made a <b>function call</b> to the give() function we wrote. Earlier, we only <b>defined</b> the <b>function</b> - that's what <b>def</b> means. It was not until the code that we just wrote that we <b>called</b> the <b>function</b> to use it.
</p>
<p>
It is important that the <b>arguments</b> line up correctly. Eric is giving the money to Lilly, so eric_money in the <b>function call</b> needs to line up with giver_amount in the <b>function definition</b>. Similarly, lilly_money needs to line up with taker_amount. Otherwise, Lilly will give Eric the money instead!
</p>
<p>
It is also just as important that we use the correct <b>variables</b> to store the values that the function returns. If you change <b>eric_money, lilly_money =</b> to <b>lilly_money, eric_money =</b> and run the code, what happens? The best description would be that Eric gave Lilly the money but then right before they left they switched wallets. This is because we returned values at the end of the function in a different order than we stored them in our main program. We took the contents of giver_amount and put it in lilly_money instead of eric_money after the flip.
</p>

In just a moment we will do an activity where you will be calling functions, but first here is the completed activity from before using our new <b>give</b> function. Notice how much easier it is to understand. Also notice that we can use <b>expressions</b> as well as <b>variables</b> for <b>function arguments</b>

In [19]:
#Starting money
eric_money = 5.0
lilly_money = 13.0
rahul_money = 7.0
ashley_money = 0.0

#Rahul gives $2 to Ashley
rahul_money, ashley_money = give(rahul_money, ashley_money, 2)

#Eric borrows all of Rahul's money
amount_borrowed = rahul_money #We need to know this value later
rahul_money, eric_money = give(rahul_money, eric_money, amount_borrowed)

#Eric spends all his money
eric_money = eric_money - eric_money

#Lilly gives half her money to Eric
lilly_money, eric_money = give(lilly_money, eric_money, lilly_money / 2.0)

#Eric gives Rahul back the same amount he borrowed earlier
eric_money, rahul_money = give(eric_money, rahul_money, amount_borrowed)

#How much money do Eric, Lilly, and Rahul have?
print "Eric has $" + str(eric_money)
print "Lilly has $" + str(lilly_money)
print "Rahul has $" + str(rahul_money)

Eric has $1.5
Lilly has $6.5
Rahul has $5.0
